From b63ed28368443febe7da63d9bc567272dd8dad98 Mon Sep 17 00:00:00 2001
From: Shanker Donthineni <sdonthineni@nvidia.com>
Date: Mon, 12 Aug 2024 22:39:25 -0500
Subject: [PATCH] NVIDIA: SAUCE: irqchip/gic-v3: Allow unused SGIs for
 drivers/modules

BugLink: https://bugs.launchpad.net/bugs/2114230

The commit 897e9e60c016 ("firmware: arm_ffa: Initial support for scheduler
receiver interrupt") adds support for SGI interrupts in the FFA driver.
However, the validation for SGIs in the GICv3 is too strict, causing the
driver probe to fail.

This patch relaxes the SGI validation check, allowing callers to use SGIs
if the requested SGI number is greater than or equal to MAX_IPI, which
fixes the TFA driver probe failure.

This issue is observed on NVIDIA server platform with FFA-v1.1.

 PTP clock support registered
 EDAC MC: Ver: 3.0.0
 ARM FF-A: Driver version 1.1
 ARM FF-A: Firmware version 1.1 found
 GICv3: [Firmware Bug]: Illegal GSI8 translation request
 ARM FF-A: Failed to create IRQ mapping!
 ARM FF-A: Notification setup failed -61, not enabled
 ARM FF-A: Failed to register driver sched callback -95
 scmi_core: SCMI protocol bus registered

This patch was sent in arm mailing list for upstream but it got
rejected.

https://patchwork.kernel.org/project/linux-arm-kernel/patch/20240813033925.925947-1-sdonthineni@nvidia.com/

The proper fix requires some kind of mechanism by which a
SGI can be requested by module but that needs discussion with arm and
it will take time. This patch will break only if MAX_IPI value gets
changed. This patch adds a BUILD_BUG_ON() to catch that situation.
Once proper solution is concluded then this patch will be reverted.

Signed-off-by: Shanker Donthineni <sdonthineni@nvidia.com>
Signed-off-by: Abhishek Sahu <abhsahu@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Noah Wager <noah.wager@canonical.com>
Acked-by: Jacob Martin <jacob.martin@canonical.com>
Signed-off--by: Brad Figg <bfigg@nvidia.com>
(backported from commit fd136cf979db)
[maskedarray: removed enum ipi_msg_type definition as it appears in
upstream commit "irqchip/gic-v5: Add GICv5 LPI/IPI support"]
Signed-off-by: Abdur Rahman <abdur.rahman@canonical.com>

(cherry picked from commit df84d5ddd125e925e912c8af1c76a71d3beb1de6 noble:linux-nvidia-6.17)
Signed-off-by: Jacob Martin <jacob.martin@canonical.com>
---
 drivers/irqchip/irq-gic-v3.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-gic-v3.c b/drivers/irqchip/irq-gic-v3.c
index 99444a1b2f..0750219ddb 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -1642,7 +1642,13 @@ static int gic_irq_domain_translate(struct irq_domain *d,
 		if(fwspec->param_count != 2)
 			return -EINVAL;
 
-		if (fwspec->param[0] < 16) {
+		/*
+		 * Below check was added on assumption that MAX_IPI
+		 * value will not be greater than 8.
+		 */
+		BUILD_BUG_ON(MAX_IPI > 8);
+
+		if (fwspec->param[0] < MAX_IPI) {
 			pr_err(FW_BUG "Illegal GSI%d translation request\n",
 			       fwspec->param[0]);
 			return -EINVAL;
From 0f52a324593fcf0655a694ac2171fc0894ac1b59 Mon Sep 17 00:00:00 2001
From: Andy Ritger <aritger@nvidia.com>
Date: Tue, 20 May 2025 17:11:58 -0700
Subject: [PATCH] NVIDIA: SAUCE: iommu/io-pgtable-arm: backport contiguous bit
 support

BugLink: https://bugs.launchpad.net/bugs/2112600

iommu/io-pgtable-arm: Support contiguous bit in translation tables

The contiguous bit in translation table entries can be used as a hint to
SMMU that a group of adjacent translation table entries have consistent
attributes and point to a contiguous and properly aligned output address
range. This enables SMMU to predict the properties of the remaining
translation table entries in the same group without accessing them. It
also allows an SMMU implementation to make more efficient use of its TLB
by using a single TLB entry to cover all translation table entries in
the same group.

In the case of 4KB granule size, there are 16 translation table entries
in one group.

This change sets the contiguous bit for such groups of entries that are
completely covered by a single call to map_pages. As it stands, the code
wouldn't set the contiguous bit if a group of adjacent descriptors is
completed by separate calls to map_pages.

Signed-off-by: Daniel Mentz <danielmentz@google.com>
Link: https://lore.kernel.org/linux-iommu/20250430231924.1481493-1-danielmentz@google.com/
[aritger: For the backport: updated the __arm_lpae_init_pte() callsite in
arm_lpae_split_blk_unmap() (arm_lpae_split_blk_unmap() is no longer
present at top of tree where the original commit was written).]
Signed-off-by: Andy Ritger <aritger@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Abdur Rahman <abdur.rahman@canonical.com>
Acked-by: Noah Wager <noah.wager@canonical.com>
Signed-off--by: Brad Figg <bfigg@nvidia.com>
(backported from commit 990fa55beac8efffb7a94b26be553d52c232b9f1 linux-nvidia-6.11)
[kobak: arm_lpae_split_blk_unmap is removed since https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=33729a5fc0caf7a97d20507acbeee6b012e7e519
iommu/io-pgtable-arm: Remove split on unmap behavior so modify as per functions]
Signed-off-by: Koba Ko <kobak@nvidia.com>
Acked-by: Matthew R. Ochs <mochs@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Jacob Martin <jacob.martin@canonical.com>
Acked-by: Abdur Rahman <abdur.rahman@canonical.com>
Signed-off-by: Ian May <ianm@nvidia.com>

(cherry picked from commit c601e605745ea9f16834da6e24049b92b759b78f noble:linux-nvidia-6.17)
Signed-off-by: Jacob Martin <jacob.martin@canonical.com>
---
 drivers/iommu/io-pgtable-arm.c | 52 ++++++++++++++++++++++++++++++++--
 1 file changed, 49 insertions(+), 3 deletions(-)

diff --git a/drivers/iommu/io-pgtable-arm.c b/drivers/iommu/io-pgtable-arm.c
index 476c0e2563..aaf2813f2c 100644
--- a/drivers/iommu/io-pgtable-arm.c
+++ b/drivers/iommu/io-pgtable-arm.c
@@ -75,6 +75,7 @@
 
 #define ARM_LPAE_PTE_NSTABLE		(((arm_lpae_iopte)1) << 63)
 #define ARM_LPAE_PTE_XN			(((arm_lpae_iopte)3) << 53)
+#define ARM_LPAE_PTE_CONT		(((arm_lpae_iopte)1) << 52)
 #define ARM_LPAE_PTE_DBM		(((arm_lpae_iopte)1) << 51)
 #define ARM_LPAE_PTE_AF			(((arm_lpae_iopte)1) << 10)
 #define ARM_LPAE_PTE_SH_NS		(((arm_lpae_iopte)0) << 8)
@@ -354,6 +355,27 @@ static void __arm_lpae_sync_pte(arm_lpae_iopte *ptep, int num_entries,
 				   sizeof(*ptep) * num_entries, DMA_TO_DEVICE);
 }
 
+static int arm_lpae_cont_ptes(int lvl, struct arm_lpae_io_pgtable *data)
+{
+	switch (ARM_LPAE_GRANULE(data)) {
+	case SZ_4K:
+		if (lvl >= 1)
+			return 16;
+		break;
+	case SZ_16K:
+		if (lvl == 2)
+			return 32;
+		else if (lvl == 3)
+			return 128;
+		break;
+	case SZ_64K:
+		if (lvl >= 2)
+			return 32;
+		break;
+	}
+	return 1;
+}
+
 static void __arm_lpae_clear_pte(arm_lpae_iopte *ptep, struct io_pgtable_cfg *cfg, int num_entries)
 {
 	for (int i = 0; i < num_entries; i++)
@@ -363,13 +385,35 @@ static void __arm_lpae_clear_pte(arm_lpae_iopte *ptep, struct io_pgtable_cfg *cf
 		__arm_lpae_sync_pte(ptep, num_entries, cfg);
 }
 
+static bool arm_lpae_use_contpte(struct arm_lpae_io_pgtable *data,
+				 unsigned long iova, phys_addr_t paddr,
+				 int lvl, int num_entries, int i)
+{
+	size_t sz = ARM_LPAE_BLOCK_SIZE(lvl, data);
+	int cont_ptes = arm_lpae_cont_ptes(lvl, data);
+	int contmask = cont_ptes - 1;
+	int contpte_addr_mask = sz * cont_ptes - 1;
+	int map_idx_start, tbl_idx;
+
+	if ((paddr & contpte_addr_mask) != (iova & contpte_addr_mask))
+		return false;
+
+	map_idx_start = ARM_LPAE_LVL_IDX(iova, lvl, data);
+	tbl_idx = map_idx_start + i;
+	if (((tbl_idx & contmask) <= i) &&
+	    (tbl_idx < ((map_idx_start + num_entries) & ~contmask)))
+		return true;
+
+	return false;
+}
+
 static size_t __arm_lpae_unmap(struct arm_lpae_io_pgtable *data,
 			       struct iommu_iotlb_gather *gather,
 			       unsigned long iova, size_t size, size_t pgcount,
 			       int lvl, arm_lpae_iopte *ptep);
 
 static void __arm_lpae_init_pte(struct arm_lpae_io_pgtable *data,
-				phys_addr_t paddr, arm_lpae_iopte prot,
+				unsigned long iova, phys_addr_t paddr, arm_lpae_iopte prot,
 				int lvl, int num_entries, arm_lpae_iopte *ptep)
 {
 	arm_lpae_iopte pte = prot;
@@ -383,7 +427,9 @@ static void __arm_lpae_init_pte(struct arm_lpae_io_pgtable *data,
 		pte |= ARM_LPAE_PTE_TYPE_BLOCK;
 
 	for (i = 0; i < num_entries; i++)
-		ptep[i] = pte | paddr_to_iopte(paddr + i * sz, data);
+		ptep[i] = pte | paddr_to_iopte(paddr + i * sz, data) |
+			  (arm_lpae_use_contpte(data, iova, paddr, lvl, num_entries, i) ?
+			  ARM_LPAE_PTE_CONT : 0);
 
 	if (!cfg->coherent_walk)
 		__arm_lpae_sync_pte(ptep, num_entries, cfg);
@@ -417,7 +463,7 @@ static int arm_lpae_init_pte(struct arm_lpae_io_pgtable *data,
 			}
 		}
 
-	__arm_lpae_init_pte(data, paddr, prot, lvl, num_entries, ptep);
+	__arm_lpae_init_pte(data, iova, paddr, prot, lvl, num_entries, ptep);
 	return 0;
 }
 
From 6a70ab5401ec8f4f69e2c9a9cc1313e7f4762ff7 Mon Sep 17 00:00:00 2001
From: Sumit Gupta <sumitg@nvidia.com>
Date: Sat, 25 Apr 2026 01:48:14 +0530
Subject: [PATCH] NVIDIA: SAUCE: cpufreq: CPPC: add autonomous mode boot
 parameter support

BugLink: https://bugs.launchpad.net/bugs/2131705

Add a kernel boot parameter 'cppc_cpufreq.auto_sel_mode' to enable
CPPC autonomous performance selection on all CPUs at system startup.
When autonomous mode is enabled, the hardware automatically adjusts
CPU performance based on workload demands using Energy Performance
Preference (EPP) hints.

When auto_sel_mode=1:
- Configure all CPUs for autonomous operation on first init
- Set EPP to performance preference (0x0)
- Use HW min/max_perf when available; otherwise initialize from caps
- Clamp desired_perf to bounds before enabling autonomous mode
- Hardware controls frequency instead of the OS governor

The boot parameter is applied only during first policy initialization.
Skip applying it on CPU hotplug to preserve runtime sysfs configuration.

This patch depends on patch [2] ("cpufreq: Set policy->min and max
as real QoS constraints") so that the policy->min/max set in
cppc_cpufreq_cpu_init() are not overridden by cpufreq_set_policy()
during init.

Reviewed-by: Randy Dunlap <rdunlap@infradead.org>
Signed-off-by: Sumit Gupta <sumitg@nvidia.com>
(backported from https://lore.kernel.org/lkml/20260424201814.230071-1-sumitg@nvidia.com/)
[jamien: hunk #2 (cppc_set_enable() insertion in cppc_cpufreq_cpu_init)
rebased onto Pierre's v2 series, which replaced the local min/max vars
with direct policy->min assignment; insertion point and code are
unchanged.]
Signed-off-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Matthew R. Ochs <mochs@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 .../admin-guide/kernel-parameters.txt         | 13 +++
 drivers/cpufreq/cppc_cpufreq.c                | 89 +++++++++++++++++--
 2 files changed, 97 insertions(+), 5 deletions(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index e8ff36982d..0363f7c18f 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -1019,6 +1019,19 @@ Kernel parameters
 			policy to use. This governor must be registered in the
 			kernel before the cpufreq driver probes.
 
+	cppc_cpufreq.auto_sel_mode=
+			[CPU_FREQ] Enable ACPI CPPC autonomous performance
+			selection. When enabled, hardware automatically adjusts
+			CPU frequency on all CPUs based on workload demands.
+			In Autonomous mode, Energy Performance Preference (EPP)
+			hints guide hardware toward performance (0x0) or energy
+			efficiency (0xff).
+			Requires ACPI CPPC autonomous selection register support.
+			Format: <bool>
+			Default: 0 (disabled)
+			0: use cpufreq governors
+			1: enable if supported by hardware
+
 	cpu_init_udelay=N
 			[X86,EARLY] Delay for N microsec between assert and de-assert
 			of APIC INIT to start processors.  This delay occurs
diff --git a/drivers/cpufreq/cppc_cpufreq.c b/drivers/cpufreq/cppc_cpufreq.c
index 8089384435..bc0c56a9a6 100644
--- a/drivers/cpufreq/cppc_cpufreq.c
+++ b/drivers/cpufreq/cppc_cpufreq.c
@@ -28,6 +28,9 @@
 
 static struct cpufreq_driver cppc_cpufreq_driver;
 
+/* Autonomous Selection boot parameter */
+static bool auto_sel_mode;
+
 #ifdef CONFIG_ACPI_CPPC_CPUFREQ_FIE
 static enum {
 	FIE_UNSET = -1,
@@ -668,6 +671,14 @@ static int cppc_cpufreq_cpu_init(struct cpufreq_policy *policy)
 	caps = &cpu_data->perf_caps;
 	policy->driver_data = cpu_data;
 
+	/*
+	 * Enable CPPC for both OS-driven and autonomous modes.
+	 * The Enable register is optional - some platforms may not support it
+	 */
+	ret = cppc_set_enable(cpu, true);
+	if (ret && ret != -EOPNOTSUPP)
+		pr_warn("Failed to enable CPPC for CPU%d (%d)\n", cpu, ret);
+
 	/*
 	 * Set min to lowest nonlinear perf to avoid any efficiency penalty (see
 	 * Section 8.4.7.1.1.5 of ACPI 6.1 spec)
@@ -720,11 +731,71 @@ static int cppc_cpufreq_cpu_init(struct cpufreq_policy *policy)
 	policy->cur = cppc_perf_to_khz(caps, caps->highest_perf);
 	cpu_data->perf_ctrls.desired_perf =  caps->highest_perf;
 
-	ret = cppc_set_perf(cpu, &cpu_data->perf_ctrls);
-	if (ret) {
-		pr_debug("Err setting perf value:%d on CPU:%d. ret:%d\n",
-			 caps->highest_perf, cpu, ret);
-		goto out;
+	/*
+	 * Enable autonomous mode on first init if boot param is set.
+	 * Check last_governor to detect first init and skip if auto_sel
+	 * is already enabled.
+	 */
+	if (auto_sel_mode && policy->last_governor[0] == '\0' &&
+	    !cpu_data->perf_ctrls.auto_sel) {
+		/* Init min/max_perf from caps if not already set by HW. */
+		if (!cpu_data->perf_ctrls.min_perf)
+			cpu_data->perf_ctrls.min_perf = caps->lowest_nonlinear_perf;
+		if (!cpu_data->perf_ctrls.max_perf)
+			cpu_data->perf_ctrls.max_perf = policy->boost_enabled ?
+				caps->highest_perf : caps->nominal_perf;
+
+		cpu_data->perf_ctrls.desired_perf =
+			clamp_t(u32, cpu_data->perf_ctrls.desired_perf,
+				cpu_data->perf_ctrls.min_perf,
+				cpu_data->perf_ctrls.max_perf);
+
+		policy->cur = cppc_perf_to_khz(caps,
+					       cpu_data->perf_ctrls.desired_perf);
+
+		/* EPP is optional - some platforms may not support it */
+		ret = cppc_set_epp(cpu, CPPC_EPP_PERFORMANCE_PREF);
+		if (ret && ret != -EOPNOTSUPP)
+			pr_warn("Failed to set EPP for CPU%d (%d)\n", cpu, ret);
+		else if (!ret)
+			cpu_data->perf_ctrls.energy_perf = CPPC_EPP_PERFORMANCE_PREF;
+
+		/* Program min/max/desired into CPPC regs before enabling auto_sel. */
+		ret = cppc_set_perf(cpu, &cpu_data->perf_ctrls);
+		if (ret) {
+			pr_debug("Err setting perf for autonomous mode CPU:%d ret:%d\n",
+				 cpu, ret);
+			goto out;
+		}
+
+		ret = cppc_set_auto_sel(cpu, true);
+		if (ret && ret != -EOPNOTSUPP) {
+			pr_warn("Failed autonomous config for CPU%d (%d)\n",
+				cpu, ret);
+			goto out;
+		}
+		if (!ret)
+			cpu_data->perf_ctrls.auto_sel = true;
+	}
+
+	if (cpu_data->perf_ctrls.auto_sel) {
+		/* Sync policy limits from HW when autonomous mode is active */
+		policy->min = cppc_perf_to_khz(caps,
+					       cpu_data->perf_ctrls.min_perf ?:
+					       caps->lowest_nonlinear_perf);
+		policy->max = cppc_perf_to_khz(caps,
+					       cpu_data->perf_ctrls.max_perf ?:
+					       (policy->boost_enabled ?
+						caps->highest_perf :
+						caps->nominal_perf));
+	} else {
+		/* Normal mode: governors control frequency */
+		ret = cppc_set_perf(cpu, &cpu_data->perf_ctrls);
+		if (ret) {
+			pr_debug("Err setting perf value:%d on CPU:%d. ret:%d\n",
+				 caps->highest_perf, cpu, ret);
+			goto out;
+		}
 	}
 
 	cppc_cpufreq_cpu_fie_init(policy);
@@ -1071,10 +1142,18 @@ static int __init cppc_cpufreq_init(void)
 
 static void __exit cppc_cpufreq_exit(void)
 {
+	unsigned int cpu;
+
+	for_each_present_cpu(cpu)
+		cppc_set_auto_sel(cpu, false);
+
 	cpufreq_unregister_driver(&cppc_cpufreq_driver);
 	cppc_freq_invariance_exit();
 }
 
+module_param(auto_sel_mode, bool, 0444);
+MODULE_PARM_DESC(auto_sel_mode, "Enable CPPC autonomous performance selection at boot");
+
 module_exit(cppc_cpufreq_exit);
 MODULE_AUTHOR("Ashwin Chaugule");
 MODULE_DESCRIPTION("CPUFreq driver based on the ACPI CPPC v5.0+ spec");
From 001906837272abbc8d210ffb79a415280705ca3e Mon Sep 17 00:00:00 2001
From: David Cemin <dcemin@nvidia.com>
Date: Wed, 26 Aug 2026 20:04:48 -0700
Subject: [PATCH] NVIDIA: SAUCE: watchdog: sbsa_gwdt: stop the watchdog across
 the whole system-sleep transition

BugLink: https://bugs.launchpad.net/bugs/2166302

The driver stops a running watchdog in its own device suspend callback
and restarts it in its resume callback. That leaves the watchdog armed,
with nobody refreshing it, for the entire early part of suspend entry:
userspace freeze, kernel thread freeze, and every device suspend
callback that runs before this device's own. The same window exists at
the tail end of resume.

When the watchdog is running from boot (early_enable=1, previously
force_enable=1 downstream; 10 s default timeout) and any device stalls
its suspend callback past the timeout, the watchdog resets the system
in the middle of suspend entry. On an arm64 laptop platform this fired
on about 7% of suspend attempts in a randomized stress run (9 resets in
124 suspends; the serial console shows the board dropping into the boot
ROM mid-entry with the watchdog reset status set in NONRST_REG2). Two
elimination runs confirm the mechanism: the identical stress matrix
with the parameter off produced zero resets in 118 suspends, and with
the first version of this change (notifier plus the original device
callbacks) applied and the watchdog force-enabled, zero resets in 198
suspends across four runs (the baseline rate predicts about 14).

Stop the watchdog from a PM notifier at the *_PREPARE events, before
tasks are frozen and device callbacks run, and restart it at the
PM_POST_* events, after everything has resumed. The driver state
(hw_armed, sleeping) lives under a lock shared with the watchdog ops:
ops->stop clears hw_armed and stops the hardware, so a userspace stop
or magic close after thaw wins over the PM_POST_* restart; ops->start
sets hw_armed but leaves the hardware stopped while sleeping is set, so
a start during the transition is deferred to PM_POST_* instead of
arming hardware nobody can refresh. The transition is recorded at
*_PREPARE whether or not the watchdog was armed at that point.

The sleep notifier is registered once, at driver init, before the
platform driver, so a probe that runs while the driver is loaded finds
a transition already recorded, whatever its interleaving with
*_PREPARE: the transition flag is global, and the probe adopts the
firmware WCS state into hw_armed, applies the flag (stopping a
firmware-started watchdog at once) and joins the instance list in one
critical section. A built-in driver therefore has no probe window at
all. What remains is the driver being loaded after PM_SUSPEND_PREPARE
has run. For that, when pm_suspend_in_progress() reports a suspend
already past task freezing, the probe records the transition itself,
under the lock, and the PM_POST_SUSPEND that always follows arms the
hardware again (the hibernation counterpart is not used: it stays true
past PM_POST_HIBERNATION and would record a transition nobody ends).
The device ->prepare callback, which the PM core runs after waiting
for outstanding probes and before any device suspend callback, applies
the same idempotent stop as the final guard for a module inserted
between *_PREPARE and the end of task freezing, or during hibernation
entry; that is the remaining exposure, and it ends before any device
suspend callback runs. No device callback re-arms the hardware:
PM_POST_* is the only restart point. A probe that fails, or an unbind,
removes the instance from the list and re-applies hw_armed to the
hardware, so a firmware-started watchdog stopped by *_PREPARE is left
as found (armed) instead of silently disabled.

Validation: the first version (notifier plus the original device
callbacks) is the one behind the 198-suspend elimination run above. The
v4 state machine (this locking and deferral, without the probe-time
adoption, the driver-init notifier and the ->prepare fallback, which a
suspend loop does not exercise) ran a further 120 suspends (60 s2idle,
60 S3, randomized order, watchdog force-enabled from boot, console
recorded through every entry) with zero resets, where the baseline rate
predicts 8 to 9. This version is compile tested (arm64, W=1, checkpatch
clean in strict mode).

This is also upstream-relevant as a companion to the early_enable
parameter: the armed-during-entry window exists for any system running
the SBSA watchdog from boot.

(backported from https://lore.kernel.org/r/20260912182107.1156221-1-dcemin@nvidia.com)
[dcemin: v8 of the posted v1, see the version paragraph above for the delta]
Signed-off-by: David Cemin <dcemin@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/watchdog/sbsa_gwdt.c | 264 ++++++++++++++++++++++++++++++-----
 1 file changed, 231 insertions(+), 33 deletions(-)

diff --git a/drivers/watchdog/sbsa_gwdt.c b/drivers/watchdog/sbsa_gwdt.c
index e04d42cc77..a2548822b1 100644
--- a/drivers/watchdog/sbsa_gwdt.c
+++ b/drivers/watchdog/sbsa_gwdt.c
@@ -43,9 +43,13 @@
 #include <linux/io.h>
 #include <linux/io-64-nonatomic-lo-hi.h>
 #include <linux/interrupt.h>
+#include <linux/list.h>
+#include <linux/mod_devicetable.h>
 #include <linux/module.h>
 #include <linux/moduleparam.h>
 #include <linux/platform_device.h>
+#include <linux/spinlock.h>
+#include <linux/suspend.h>
 #include <linux/uaccess.h>
 #include <linux/watchdog.h>
 #include <asm/arch_timer.h>
@@ -87,6 +91,10 @@
  *			indicate whether to adjust wdd->timeout to avoid a race with WS0
  * @refresh_base:	Virtual address of the watchdog refresh frame
  * @control_base:	Virtual address of the watchdog control frame
+ * @node:		Entry in the list of instances the sleep notifier owns
+ * @hw_armed:		The watchdog is logically running (started by firmware,
+ *			early_enable or userspace); the hardware follows it except
+ *			while a system sleep transition is in progress
  */
 struct sbsa_gwdt {
 	struct watchdog_device	wdd;
@@ -95,8 +103,20 @@ struct sbsa_gwdt {
 	bool			need_ws0_race_workaround;
 	void __iomem		*refresh_base;
 	void __iomem		*control_base;
+	struct list_head	node;
+	bool			hw_armed;
 };
 
+/*
+ * System sleep state shared by all instances. The notifier recording a
+ * transition is registered once, at driver init, before any device can be
+ * probed, so a probe at any later point finds the transition already
+ * recorded instead of racing its own registration against *_PREPARE.
+ */
+static LIST_HEAD(sbsa_gwdt_list);
+static DEFINE_SPINLOCK(sbsa_gwdt_lock); /* list, sleeping, hw_armed */
+static bool sbsa_gwdt_sleeping;
+
 #define DEFAULT_TIMEOUT		10 /* seconds */
 
 static unsigned int timeout;
@@ -244,12 +264,33 @@ static void sbsa_gwdt_get_version(struct watchdog_device *wdd)
 		!action && (impl == SBSA_GWDT_IMPL_MEDIATEK);
 }
 
+static void sbsa_gwdt_hw_start(struct sbsa_gwdt *gwdt)
+{
+	/* writing WCS will cause an explicit watchdog refresh */
+	writel(SBSA_GWDT_WCS_EN, gwdt->control_base + SBSA_GWDT_WCS);
+}
+
+static void sbsa_gwdt_hw_stop(struct sbsa_gwdt *gwdt)
+{
+	/* Simply write 0 to WCS to clean WCS_EN bit */
+	writel(0, gwdt->control_base + SBSA_GWDT_WCS);
+}
+
 static int sbsa_gwdt_start(struct watchdog_device *wdd)
 {
 	struct sbsa_gwdt *gwdt = watchdog_get_drvdata(wdd);
+	unsigned long flags;
 
-	/* writing WCS will cause an explicit watchdog refresh */
-	writel(SBSA_GWDT_WCS_EN, gwdt->control_base + SBSA_GWDT_WCS);
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	gwdt->hw_armed = true;
+	/*
+	 * While a system sleep transition is in progress nobody can refresh
+	 * the watchdog: leave the hardware stopped and let the PM_POST_*
+	 * notifier arm it once everything has resumed.
+	 */
+	if (!sbsa_gwdt_sleeping)
+		sbsa_gwdt_hw_start(gwdt);
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
 
 	return 0;
 }
@@ -257,9 +298,12 @@ static int sbsa_gwdt_start(struct watchdog_device *wdd)
 static int sbsa_gwdt_stop(struct watchdog_device *wdd)
 {
 	struct sbsa_gwdt *gwdt = watchdog_get_drvdata(wdd);
+	unsigned long flags;
 
-	/* Simply write 0 to WCS to clean WCS_EN bit */
-	writel(0, gwdt->control_base + SBSA_GWDT_WCS);
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	gwdt->hw_armed = false;
+	sbsa_gwdt_hw_stop(gwdt);
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
 
 	return 0;
 }
@@ -288,12 +332,133 @@ static const struct watchdog_ops sbsa_gwdt_ops = {
 	.get_timeleft	= sbsa_gwdt_get_timeleft,
 };
 
+/*
+ * Per-device suspend/resume callbacks alone would stop the watchdog only
+ * once this device itself is suspended, one of the last steps of suspend
+ * entry, and restart it during device resume, before tasks are thawed. A
+ * watchdog running from boot (early_enable) would therefore be armed, with
+ * nobody refreshing it, through task freezing and every other device's
+ * suspend callback on the way down, and again from device resume until
+ * userspace runs on the way up; anything stalling past the timeout in
+ * either window resets the system.
+ *
+ * Own the transition from a PM notifier instead: stop at the *_PREPARE
+ * events, before anything is frozen, and restart at PM_POST_*, after
+ * everything has resumed. The state (hw_armed per instance, one sleeping
+ * flag) is kept under a lock shared with the watchdog ops so that a
+ * userspace stop or magic close after thaw cannot race the restart, and a
+ * start requested while the transition is in progress is deferred to
+ * PM_POST_*. The transition is recorded at *_PREPARE whether or not the
+ * watchdog was armed at that moment, so a start between *_PREPARE and
+ * task freezing is deferred as well instead of arming hardware nobody
+ * refreshes.
+ *
+ * The notifier is registered at driver init, so a probe that runs while
+ * the driver is loaded finds a transition already recorded, whatever its
+ * interleaving with *_PREPARE, and applies it when it adopts the hardware
+ * state. What remains is the driver being loaded after PM_SUSPEND_PREPARE
+ * has run: the probe then checks for a suspend already past task freezing
+ * and stops the hardware itself, and the device ->prepare callback, which
+ * the PM core runs after waiting for outstanding probes and before any
+ * device suspend callback, applies the same stop. Neither has a resume
+ * counterpart: PM_POST_* is the only place that re-arms the hardware.
+ */
+static void sbsa_gwdt_sleep_stop(void)
+{
+	struct sbsa_gwdt *gwdt;
+	unsigned long flags;
+
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	/*
+	 * Idempotent on purpose: the *_PREPARE notifier and the device
+	 * ->prepare callback both land here, and a probe that adopted a
+	 * firmware-armed watchdog after *_PREPARE relies on a later call
+	 * actually stopping the hardware.
+	 */
+	sbsa_gwdt_sleeping = true;
+	list_for_each_entry(gwdt, &sbsa_gwdt_list, node)
+		if (gwdt->hw_armed)
+			sbsa_gwdt_hw_stop(gwdt);
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
+}
+
+static void sbsa_gwdt_sleep_restart(void)
+{
+	struct sbsa_gwdt *gwdt;
+	unsigned long flags;
+
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	if (sbsa_gwdt_sleeping) {
+		sbsa_gwdt_sleeping = false;
+		list_for_each_entry(gwdt, &sbsa_gwdt_list, node)
+			if (gwdt->hw_armed)
+				sbsa_gwdt_hw_start(gwdt);
+	}
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
+}
+
+static int sbsa_gwdt_pm_notify(struct notifier_block *nb, unsigned long mode,
+			       void *data)
+{
+	switch (mode) {
+	case PM_SUSPEND_PREPARE:
+	case PM_HIBERNATION_PREPARE:
+	case PM_RESTORE_PREPARE:
+		sbsa_gwdt_sleep_stop();
+		break;
+	case PM_POST_SUSPEND:
+	case PM_POST_HIBERNATION:
+	case PM_POST_RESTORE:
+		sbsa_gwdt_sleep_restart();
+		break;
+	}
+
+	return NOTIFY_DONE;
+}
+
+static struct notifier_block sbsa_gwdt_pm_nb = {
+	.notifier_call = sbsa_gwdt_pm_notify,
+};
+
+static void sbsa_gwdt_remove_instance(void *data)
+{
+	struct sbsa_gwdt *gwdt = data;
+	unsigned long flags;
+
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	list_del(&gwdt->node);
+	/*
+	 * PM_POST_* no longer applies to this instance. If a transition
+	 * stopped the hardware, put it back into the state the last start or
+	 * stop asked for, so a probe that fails while the system heads into
+	 * sleep does not leave a firmware-started watchdog silently disabled;
+	 * without a driver such a watchdog is armed and unrefreshed whether
+	 * or not a transition is in progress, exactly as a probe failure
+	 * outside a transition leaves it.
+	 */
+	if (sbsa_gwdt_sleeping && gwdt->hw_armed)
+		sbsa_gwdt_hw_start(gwdt);
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
+}
+
+static int sbsa_gwdt_prepare(struct device *dev)
+{
+	sbsa_gwdt_sleep_stop();
+
+	return 0;
+}
+
+static const struct dev_pm_ops sbsa_gwdt_pm_ops = {
+	.prepare = pm_sleep_ptr(sbsa_gwdt_prepare),
+};
+
 static int sbsa_gwdt_probe(struct platform_device *pdev)
 {
 	void __iomem *rf_base, *cf_base;
 	struct device *dev = &pdev->dev;
 	struct watchdog_device *wdd;
 	struct sbsa_gwdt *gwdt;
+	unsigned long flags;
 	int ret, irq;
 	u32 status;
 
@@ -318,6 +483,34 @@ static int sbsa_gwdt_probe(struct platform_device *pdev)
 	gwdt->clk = arch_timer_get_cntfrq();
 	gwdt->refresh_base = rf_base;
 	gwdt->control_base = cf_base;
+	status = readl(cf_base + SBSA_GWDT_WCS);
+
+	/*
+	 * Adopt the firmware state and join the list under the lock: a
+	 * transition the notifier has already recorded applies to this
+	 * instance at once, and a *_PREPARE landing after this point finds
+	 * hw_armed set. pm_suspend_in_progress() covers the driver being
+	 * loaded after PM_SUSPEND_PREPARE ran: it is true from the end of
+	 * task freezing until the devices have resumed, and PM_POST_SUSPEND
+	 * always follows its clearing, so a transition seen here is one the
+	 * notifier will see the end of; read under the lock, a PM_POST_SUSPEND
+	 * racing it finds the flag set. The hibernation counterpart stays
+	 * true past PM_POST_HIBERNATION and would record a transition nobody
+	 * ends, so that case, like the interval between *_PREPARE and the end
+	 * of task freezing, relies on the device ->prepare callback instead.
+	 */
+	spin_lock_irqsave(&sbsa_gwdt_lock, flags);
+	gwdt->hw_armed = !!(status & SBSA_GWDT_WCS_EN);
+	if (pm_suspend_in_progress())
+		sbsa_gwdt_sleeping = true;
+	if (sbsa_gwdt_sleeping && gwdt->hw_armed)
+		sbsa_gwdt_hw_stop(gwdt);
+	list_add(&gwdt->node, &sbsa_gwdt_list);
+	spin_unlock_irqrestore(&sbsa_gwdt_lock, flags);
+
+	ret = devm_add_action_or_reset(dev, sbsa_gwdt_remove_instance, gwdt);
+	if (ret)
+		return ret;
 
 	wdd = &gwdt->wdd;
 	wdd->parent = dev;
@@ -342,11 +535,15 @@ static int sbsa_gwdt_probe(struct platform_device *pdev)
 		wdd->min_timeout = 3;
 	}
 
-	status = readl(cf_base + SBSA_GWDT_WCS);
 	if (status & SBSA_GWDT_WCS_WS1) {
 		dev_warn(dev, "System reset by WDT.\n");
 		wdd->bootstatus |= WDIOF_CARDRESET;
 	}
+	/*
+	 * hw_armed already reflects WCS_EN (adopted above, under the lock);
+	 * a transition recorded before or since has stopped the hardware
+	 * and PM_POST_* re-arms it.
+	 */
 	if (status & SBSA_GWDT_WCS_EN)
 		set_bit(WDOG_HW_RUNNING, &wdd->status);
 
@@ -398,32 +595,6 @@ static int sbsa_gwdt_probe(struct platform_device *pdev)
 	return 0;
 }
 
-/* Disable watchdog if it is active during suspend */
-static int __maybe_unused sbsa_gwdt_suspend(struct device *dev)
-{
-	struct sbsa_gwdt *gwdt = dev_get_drvdata(dev);
-
-	if (watchdog_hw_running(&gwdt->wdd))
-		sbsa_gwdt_stop(&gwdt->wdd);
-
-	return 0;
-}
-
-/* Enable watchdog if necessary */
-static int __maybe_unused sbsa_gwdt_resume(struct device *dev)
-{
-	struct sbsa_gwdt *gwdt = dev_get_drvdata(dev);
-
-	if (watchdog_hw_running(&gwdt->wdd))
-		sbsa_gwdt_start(&gwdt->wdd);
-
-	return 0;
-}
-
-static const struct dev_pm_ops sbsa_gwdt_pm_ops = {
-	SET_SYSTEM_SLEEP_PM_OPS(sbsa_gwdt_suspend, sbsa_gwdt_resume)
-};
-
 static const struct of_device_id sbsa_gwdt_of_match[] = {
 	{ .compatible = "arm,sbsa-gwdt", },
 	{},
@@ -439,14 +610,41 @@ MODULE_DEVICE_TABLE(platform, sbsa_gwdt_pdev_match);
 static struct platform_driver sbsa_gwdt_driver = {
 	.driver = {
 		.name = DRV_NAME,
-		.pm = &sbsa_gwdt_pm_ops,
+		.pm = pm_sleep_ptr(&sbsa_gwdt_pm_ops),
 		.of_match_table = sbsa_gwdt_of_match,
 	},
 	.probe = sbsa_gwdt_probe,
 	.id_table = sbsa_gwdt_pdev_match,
 };
 
-module_platform_driver(sbsa_gwdt_driver);
+static int __init sbsa_gwdt_init(void)
+{
+	int ret;
+
+	/*
+	 * Register the sleep notifier before any device can be probed, so
+	 * that every transition from here on is recorded before a probe
+	 * adopts a running watchdog. Its failure is fatal: without it a
+	 * running watchdog would survive into system sleep unrefreshed.
+	 */
+	ret = register_pm_notifier(&sbsa_gwdt_pm_nb);
+	if (ret)
+		return ret;
+
+	ret = platform_driver_register(&sbsa_gwdt_driver);
+	if (ret)
+		unregister_pm_notifier(&sbsa_gwdt_pm_nb);
+
+	return ret;
+}
+module_init(sbsa_gwdt_init);
+
+static void __exit sbsa_gwdt_exit(void)
+{
+	platform_driver_unregister(&sbsa_gwdt_driver);
+	unregister_pm_notifier(&sbsa_gwdt_pm_nb);
+}
+module_exit(sbsa_gwdt_exit);
 
 MODULE_DESCRIPTION("SBSA Generic Watchdog Driver");
 MODULE_AUTHOR("Fu Wei <fu.wei@linaro.org>");
From e95f46e4c0e86d36c5c195e5707c4c649c90c955 Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Sat, 6 Jun 2026 08:51:52 -0700
Subject: [PATCH] NVIDIA: SAUCE: pmdomain: Store cached power-down state index
 as an unsigned int

BugLink: https://bugs.launchpad.net/bugs/2167301

genpd caches the state index selected by the governor so the next
power-down decision can reuse it while constraints remain unchanged.

The cache currently stores the index in a bool, truncating every index
above 1. Use unsigned int to match generic_pm_domain.state_idx and preserve
the complete selected index. Domains that only select state 0 or 1 are
unchanged.

Fixes: e94999688e3a ("PM / Domains: Add genpd governor for CPUs")
Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Reviewed-by: Ishan Shah <ishah@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 include/linux/pm_domain.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index f925614aeb..103e74486e 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -171,7 +171,7 @@ struct genpd_governor_data {
 	ktime_t last_enter;
 	bool reflect_residency;
 	bool cached_power_down_ok;
-	bool cached_power_down_state_idx;
+	unsigned int cached_power_down_state_idx;
 };
 
 struct genpd_power_state {
From f2f87aa6aa27401510455298e9fcbea2df78bc35 Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Mon, 29 Jun 2026 21:36:09 -0700
Subject: [PATCH] NVIDIA: SAUCE: sched: idle: Treat wakeup no-constraint as
 unbounded

BugLink: https://bugs.launchpad.net/bugs/2167301

The CPU wakeup latency QoS no-constraint value is a sentinel, not a finite
latency budget.  When entering s2idle, use U64_MAX for cpuidle state
selection if no CPU wakeup latency constraint is active.

This preserves finite constraints while avoiding rejection of states with
firmware-provided latency values larger than S32_MAX usec.

Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 kernel/sched/idle.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/idle.c b/kernel/sched/idle.c
index 052435f4d3..056245ae17 100644
--- a/kernel/sched/idle.c
+++ b/kernel/sched/idle.c
@@ -7,6 +7,7 @@
  *        tasks which are handled in sched/fair.c )
  */
 #include <linux/cpuidle.h>
+#include <linux/pm_qos.h>
 #include <linux/suspend.h>
 #include <linux/livepatch.h>
 #include "sched.h"
@@ -214,8 +215,14 @@ static void cpuidle_idle_call(bool stop_tick)
 		u64 max_latency_ns;
 
 		if (idle_should_enter_s2idle()) {
-			max_latency_ns = cpu_wakeup_latency_qos_limit() *
-					 NSEC_PER_USEC;
+			s32 latency_limit_us = cpu_wakeup_latency_qos_limit();
+
+			if (latency_limit_us ==
+			    PM_QOS_RESUME_LATENCY_NO_CONSTRAINT)
+				max_latency_ns = U64_MAX;
+			else
+				max_latency_ns = (u64)latency_limit_us *
+						 NSEC_PER_USEC;
 
 			entered_state = call_cpuidle_s2idle(drv, dev,
 							    max_latency_ns);
From b6667ee782222e44e85d9f3f662d3bc9900d4852 Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Tue, 1 Sep 2026 22:48:30 +0000
Subject: [PATCH] NVIDIA: SAUCE: pmdomain: Treat unconstrained CPU wake latency
 as unlimited

BugLink: https://bugs.launchpad.net/bugs/2167301

cpu_system_power_down_ok() converts the CPU resume-latency QoS value to
nanoseconds before comparing it with genpd state latency. However,
PM_QOS_RESUME_LATENCY_NO_CONSTRAINT is a sentinel rather than a finite
microsecond limit. Treating it as an ordinary value can reject a valid
state whose combined latency is larger than the converted sentinel.

Detect the sentinel before conversion and allow any state latency when no
constraint is present. Continue enforcing zero and finite latency requests
normally.

Fixes: 8e7de6dc4209 ("pmdomain: Respect the CPU system wakeup QoS limit for s2idle")
Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Reviewed-by: Ishan Shah <ishah@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/pmdomain/governor.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/pmdomain/governor.c b/drivers/pmdomain/governor.c
index 96737abbb4..fc86190ca8 100644
--- a/drivers/pmdomain/governor.c
+++ b/drivers/pmdomain/governor.c
@@ -427,8 +427,10 @@ static bool cpu_power_down_ok(struct dev_pm_domain *pd)
 
 static bool cpu_system_power_down_ok(struct dev_pm_domain *pd)
 {
-	s64 constraint_ns = cpu_wakeup_latency_qos_limit() * NSEC_PER_USEC;
 	struct generic_pm_domain *genpd = pd_to_genpd(pd);
+	s32 constraint_us = cpu_wakeup_latency_qos_limit();
+	bool no_constraint;
+	s64 constraint_ns;
 	int state_idx = genpd->state_count - 1;
 
 	if (!(genpd->flags & GENPD_FLAG_CPU_DOMAIN)) {
@@ -436,12 +438,17 @@ static bool cpu_system_power_down_ok(struct dev_pm_domain *pd)
 		return true;
 	}
 
+	no_constraint = constraint_us == PM_QOS_RESUME_LATENCY_NO_CONSTRAINT;
+	constraint_ns = no_constraint ? PM_QOS_RESUME_LATENCY_NO_CONSTRAINT_NS :
+			(s64)constraint_us * NSEC_PER_USEC;
+
 	/* Find the deepest state for the latency constraint. */
 	while (state_idx >= 0) {
 		s64 latency_ns = genpd->states[state_idx].power_off_latency_ns +
 				 genpd->states[state_idx].power_on_latency_ns;
 
-		if (latency_ns <= constraint_ns) {
+		/* The no-constraint value is a sentinel, not a latency budget. */
+		if (no_constraint || latency_ns <= constraint_ns) {
 			genpd->state_idx = state_idx;
 			return true;
 		}
From e67b10d0eb72bb9885b8b6d953c12b7db7e31ec7 Mon Sep 17 00:00:00 2001
From: Ishan Shah <ishah@nvidia.com>
Date: Fri, 4 Sep 2026 21:11:55 +0000
Subject: [PATCH] NVIDIA: SAUCE: pmdomain: Correct rejected idle-state
 accounting

BugLink: https://bugs.launchpad.net/bugs/2167301

pm_genpd_inc_rejected() accepts the state index whose asynchronous
power-off attempt failed, but updates the domain current state index
instead. The current index may have changed before the correction runs,
causing the wrong usage and rejected counters to be adjusted.

The helper also leaves usage_s2idle incremented when a suspend-to-idle
entry fails. Accept the entry context from callers, use the supplied index,
and undo the s2idle counter when genpd incremented it. Reject an
out-of-range index before accessing the state table.

Fixes: 9861f21ff16b ("pmdomain: core: Add genpd helper to correct the usage/rejected counters")
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/cpuidle/cpuidle-psci.c |  2 +-
 drivers/pmdomain/core.c        | 14 +++++++++++---
 include/linux/pm_domain.h      |  4 ++--
 3 files changed, 14 insertions(+), 6 deletions(-)

diff --git a/drivers/cpuidle/cpuidle-psci.c b/drivers/cpuidle/cpuidle-psci.c
index b250d0dde7..009bbedcd9 100644
--- a/drivers/cpuidle/cpuidle-psci.c
+++ b/drivers/cpuidle/cpuidle-psci.c
@@ -99,7 +99,7 @@ static __cpuidle int __psci_enter_domain_idle_state(struct cpuidle_device *dev,
 
 	/* Correct domain-idlestate statistics if we failed to enter. */
 	if (ret == -1 && ds->state)
-		pm_genpd_inc_rejected(ds->pd, ds->state_idx);
+		pm_genpd_inc_rejected(ds->pd, ds->state_idx, s2idle);
 
 	/* Clear the domain state to start fresh when back from idle. */
 	psci_clear_domain_state();
diff --git a/drivers/pmdomain/core.c b/drivers/pmdomain/core.c
index 842c4169e2..51ed2b4a8e 100644
--- a/drivers/pmdomain/core.c
+++ b/drivers/pmdomain/core.c
@@ -803,6 +803,7 @@ EXPORT_SYMBOL_GPL(dev_pm_genpd_is_on);
  *
  * @genpd: The PM domain the idle-state belongs to.
  * @state_idx: The index of the idle-state that failed.
+ * @s2idle: Whether the failed attempt was made for suspend-to-idle.
  *
  * In some special cases the ->power_off() callback is asynchronously powering
  * off the PM domain, leading to that it may return zero to indicate success,
@@ -814,11 +815,18 @@ EXPORT_SYMBOL_GPL(dev_pm_genpd_is_on);
  * while this routine is getting called.
  */
 void pm_genpd_inc_rejected(struct generic_pm_domain *genpd,
-			   unsigned int state_idx)
+			   unsigned int state_idx, bool s2idle)
 {
 	genpd_lock(genpd);
-	genpd->states[genpd->state_idx].rejected++;
-	genpd->states[genpd->state_idx].usage--;
+	if (WARN_ON_ONCE(state_idx >= genpd->state_count))
+		goto out;
+
+	genpd->states[state_idx].rejected++;
+	genpd->states[state_idx].usage--;
+	if (s2idle && genpd->gov && genpd->gov->system_power_down_ok)
+		genpd->states[state_idx].usage_s2idle--;
+
+out:
 	genpd_unlock(genpd);
 }
 EXPORT_SYMBOL_GPL(pm_genpd_inc_rejected);
diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index 103e74486e..0208d6c85d 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -319,7 +319,7 @@ int pm_genpd_init(struct generic_pm_domain *genpd,
 		  struct dev_power_governor *gov, bool is_off);
 int pm_genpd_remove(struct generic_pm_domain *genpd);
 void pm_genpd_inc_rejected(struct generic_pm_domain *genpd,
-			   unsigned int state_idx);
+			   unsigned int state_idx, bool s2idle);
 struct device *dev_to_genpd_dev(struct device *dev);
 int dev_pm_genpd_set_performance_state(struct device *dev, unsigned int state);
 int dev_pm_genpd_add_notifier(struct device *dev, struct notifier_block *nb);
@@ -373,7 +373,7 @@ static inline int pm_genpd_remove(struct generic_pm_domain *genpd)
 }
 
 static inline void pm_genpd_inc_rejected(struct generic_pm_domain *genpd,
-					 unsigned int state_idx)
+					 unsigned int state_idx, bool s2idle)
 { }
 
 static inline struct device *dev_to_genpd_dev(struct device *dev)
From cc2be96ae537fdd0eb6211a0fa3065b5e1444e61 Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Tue, 15 Sep 2026 18:04:27 +0000
Subject: [PATCH] NVIDIA: SAUCE: pmdomain: Distinguish s2idle state usage

BugLink: https://bugs.launchpad.net/bugs/2167301

Synchronous genpd power-off is used for more than suspend-to-idle.
System-sleep and PREEMPT_RT CPU-hotplug paths also use
dev_pm_genpd_suspend(), so inferring the accounting context from the
presence of a system_power_down_ok() callback classifies those transitions
as s2idle.

Add an explicit s2idle variant of the suspend helper and propagate that
context through synchronous parent power-off. Keep the existing helper as
the non-s2idle path, update the PSCI and RISC-V SBI s2idle callers, and use
the same explicit context when correcting a rejected selection.

Co-developed-by: Ishan Shah <ishah@nvidia.com>
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/cpuidle/cpuidle-psci.c      |  2 +-
 drivers/cpuidle/cpuidle-riscv-sbi.c |  2 +-
 drivers/pmdomain/core.c             | 41 ++++++++++++++++++-----------
 include/linux/pm_domain.h           |  2 ++
 4 files changed, 29 insertions(+), 18 deletions(-)

diff --git a/drivers/cpuidle/cpuidle-psci.c b/drivers/cpuidle/cpuidle-psci.c
index 009bbedcd9..1562a5b09d 100644
--- a/drivers/cpuidle/cpuidle-psci.c
+++ b/drivers/cpuidle/cpuidle-psci.c
@@ -78,7 +78,7 @@ static __cpuidle int __psci_enter_domain_idle_state(struct cpuidle_device *dev,
 
 	/* Do runtime PM to manage a hierarchical CPU toplogy. */
 	if (s2idle)
-		dev_pm_genpd_suspend(pd_dev);
+		dev_pm_genpd_suspend_s2idle(pd_dev);
 	else
 		pm_runtime_put_sync_suspend(pd_dev);
 
diff --git a/drivers/cpuidle/cpuidle-riscv-sbi.c b/drivers/cpuidle/cpuidle-riscv-sbi.c
index a13e90bb43..a84085a285 100644
--- a/drivers/cpuidle/cpuidle-riscv-sbi.c
+++ b/drivers/cpuidle/cpuidle-riscv-sbi.c
@@ -104,7 +104,7 @@ static __cpuidle int __sbi_enter_domain_idle_state(struct cpuidle_device *dev,
 
 	/* Do runtime PM to manage a hierarchical CPU toplogy. */
 	if (s2idle)
-		dev_pm_genpd_suspend(pd_dev);
+		dev_pm_genpd_suspend_s2idle(pd_dev);
 	else
 		pm_runtime_put_sync_suspend(pd_dev);
 
diff --git a/drivers/pmdomain/core.c b/drivers/pmdomain/core.c
index 51ed2b4a8e..230c7f34a1 100644
--- a/drivers/pmdomain/core.c
+++ b/drivers/pmdomain/core.c
@@ -823,7 +823,7 @@ void pm_genpd_inc_rejected(struct generic_pm_domain *genpd,
 
 	genpd->states[state_idx].rejected++;
 	genpd->states[state_idx].usage--;
-	if (s2idle && genpd->gov && genpd->gov->system_power_down_ok)
+	if (s2idle)
 		genpd->states[state_idx].usage_s2idle--;
 
 out:
@@ -1400,6 +1400,7 @@ late_initcall_sync(genpd_power_off_unused);
  * genpd_sync_power_off - Synchronously power off a PM domain and its parents.
  * @genpd: PM domain to power off, if possible.
  * @use_lock: use the lock.
+ * @s2idle: account successful state selection as suspend-to-idle.
  * @depth: nesting count for lockdep.
  *
  * Check if the given PM domain can be powered off (during system suspend or
@@ -1410,7 +1411,7 @@ late_initcall_sync(genpd_power_off_unused);
  * these cases the lock must be held.
  */
 static void genpd_sync_power_off(struct generic_pm_domain *genpd, bool use_lock,
-				 unsigned int depth)
+				 bool s2idle, unsigned int depth)
 {
 	struct gpd_link *link;
 
@@ -1442,11 +1443,7 @@ static void genpd_sync_power_off(struct generic_pm_domain *genpd, bool use_lock,
 	} else {
 		genpd->states[genpd->state_idx].usage++;
 
-		/*
-		 * The ->system_power_down_ok() callback is currently used only
-		 * for s2idle. Use it to know when to update the usage counter.
-		 */
-		if (genpd->gov && genpd->gov->system_power_down_ok)
+		if (s2idle)
 			genpd->states[genpd->state_idx].usage_s2idle++;
 	}
 
@@ -1458,7 +1455,8 @@ static void genpd_sync_power_off(struct generic_pm_domain *genpd, bool use_lock,
 		if (use_lock)
 			genpd_lock_nested(link->parent, depth + 1);
 
-		genpd_sync_power_off(link->parent, use_lock, depth + 1);
+		genpd_sync_power_off(link->parent, use_lock, s2idle,
+				     depth + 1);
 
 		if (use_lock)
 			genpd_unlock(link->parent);
@@ -1576,7 +1574,7 @@ static int genpd_finish_suspend(struct device *dev,
 
 	genpd_lock(genpd);
 	genpd->suspended_count++;
-	genpd_sync_power_off(genpd, true, 0);
+	genpd_sync_power_off(genpd, true, false, 0);
 	genpd_unlock(genpd);
 
 	return 0;
@@ -1742,7 +1740,7 @@ static void genpd_complete(struct device *dev)
 	genpd_unlock(genpd);
 }
 
-static void genpd_switch_state(struct device *dev, bool suspend)
+static void genpd_switch_state(struct device *dev, bool suspend, bool s2idle)
 {
 	struct generic_pm_domain *genpd;
 	bool use_lock;
@@ -1758,7 +1756,7 @@ static void genpd_switch_state(struct device *dev, bool suspend)
 
 	if (suspend) {
 		genpd->suspended_count++;
-		genpd_sync_power_off(genpd, use_lock, 0);
+		genpd_sync_power_off(genpd, use_lock, s2idle, 0);
 	} else {
 		genpd_sync_power_on(genpd, use_lock, 0);
 		genpd->suspended_count--;
@@ -1773,16 +1771,27 @@ static void genpd_switch_state(struct device *dev, bool suspend)
  * @dev: The device that is attached to the genpd, that can be suspended.
  *
  * This routine should typically be called for a device that needs to be
- * suspended during the syscore suspend phase. It may also be called during
- * suspend-to-idle to suspend a corresponding CPU device that is attached to a
- * genpd.
+ * suspended during the syscore suspend phase.
  */
 void dev_pm_genpd_suspend(struct device *dev)
 {
-	genpd_switch_state(dev, true);
+	genpd_switch_state(dev, true, false);
 }
 EXPORT_SYMBOL_GPL(dev_pm_genpd_suspend);
 
+/**
+ * dev_pm_genpd_suspend_s2idle - Suspend the genpd for @dev during s2idle
+ * @dev: The device that is attached to the genpd, that can be suspended.
+ *
+ * This is equivalent to dev_pm_genpd_suspend(), but records successful domain
+ * state selections as suspend-to-idle entries.
+ */
+void dev_pm_genpd_suspend_s2idle(struct device *dev)
+{
+	genpd_switch_state(dev, true, true);
+}
+EXPORT_SYMBOL_GPL(dev_pm_genpd_suspend_s2idle);
+
 /**
  * dev_pm_genpd_resume - Synchronously try to resume the genpd for @dev
  * @dev: The device that is attached to the genpd, which needs to be resumed.
@@ -1793,7 +1802,7 @@ EXPORT_SYMBOL_GPL(dev_pm_genpd_suspend);
  */
 void dev_pm_genpd_resume(struct device *dev)
 {
-	genpd_switch_state(dev, false);
+	genpd_switch_state(dev, false, false);
 }
 EXPORT_SYMBOL_GPL(dev_pm_genpd_resume);
 
diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index 0208d6c85d..4ac6005610 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -434,9 +434,11 @@ static inline bool dev_pm_genpd_is_on(struct device *dev)
 
 #ifdef CONFIG_PM_GENERIC_DOMAINS_SLEEP
 void dev_pm_genpd_suspend(struct device *dev);
+void dev_pm_genpd_suspend_s2idle(struct device *dev);
 void dev_pm_genpd_resume(struct device *dev);
 #else
 static inline void dev_pm_genpd_suspend(struct device *dev) {}
+static inline void dev_pm_genpd_suspend_s2idle(struct device *dev) {}
 static inline void dev_pm_genpd_resume(struct device *dev) {}
 #endif
 
From 5231c8fe672a6e8e7cde1b6cced64c2ba5917c14 Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Tue, 15 Sep 2026 18:05:26 +0000
Subject: [PATCH] NVIDIA: SAUCE: cpuidle: psci: Correct rejected states for the
 full domain path

BugLink: https://bugs.launchpad.net/bugs/2167301

Each successful genpd power-off callback replaces the single per-CPU PSCI
domain-state record. In a multilevel topology, firmware rejection can
therefore correct the usage and rejected counters for only the final
selected domain.

Count the attached domain paths during CPU initialization and preallocate
enough per-CPU storage for every selection. Collect each selected domain
only while preparing a cpuidle firmware entry, use the final record for the
composite PSCI state, and correct every recorded domain if firmware rejects
it. Propagate an exhausted record array back through the genpd power-off
callback.

Keeping collection scoped to the cpuidle transaction also prevents syscore
and CPU-hotplug genpd transitions from accumulating irrelevant state.

Co-developed-by: Ishan Shah <ishah@nvidia.com>
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/cpuidle/cpuidle-psci-domain.c |  4 +-
 drivers/cpuidle/cpuidle-psci.c        | 92 +++++++++++++++++++++++----
 drivers/cpuidle/cpuidle-psci.h        |  4 +-
 3 files changed, 84 insertions(+), 16 deletions(-)

diff --git a/drivers/cpuidle/cpuidle-psci-domain.c b/drivers/cpuidle/cpuidle-psci-domain.c
index b9e4ad7d43..c4984718f1 100644
--- a/drivers/cpuidle/cpuidle-psci-domain.c
+++ b/drivers/cpuidle/cpuidle-psci-domain.c
@@ -39,9 +39,7 @@ static int psci_pd_power_off(struct generic_pm_domain *pd)
 
 	/* OSI mode is enabled, set the corresponding domain state. */
 	pd_state = state->data;
-	psci_set_domain_state(pd, pd->state_idx, *pd_state);
-
-	return 0;
+	return psci_set_domain_state(pd, pd->state_idx, *pd_state);
 }
 
 static int psci_pd_init(struct device_node *np, bool use_osi)
diff --git a/drivers/cpuidle/cpuidle-psci.c b/drivers/cpuidle/cpuidle-psci.c
index 1562a5b09d..3d45edf749 100644
--- a/drivers/cpuidle/cpuidle-psci.c
+++ b/drivers/cpuidle/cpuidle-psci.c
@@ -34,6 +34,10 @@
 struct psci_cpuidle_data {
 	u32 *psci_states;
 	struct device *dev;
+	struct psci_cpuidle_domain_state *domain_states;
+	unsigned int domain_state_count;
+	unsigned int domain_state_capacity;
+	bool collect_domain_states;
 };
 
 struct psci_cpuidle_domain_state {
@@ -43,22 +47,32 @@ struct psci_cpuidle_domain_state {
 };
 
 static DEFINE_PER_CPU_READ_MOSTLY(struct psci_cpuidle_data, psci_cpuidle_data);
-static DEFINE_PER_CPU(struct psci_cpuidle_domain_state, psci_domain_state);
 static bool psci_cpuidle_use_syscore;
 
-void psci_set_domain_state(struct generic_pm_domain *pd, unsigned int state_idx,
-			   u32 state)
+int psci_set_domain_state(struct generic_pm_domain *pd, unsigned int state_idx,
+			  u32 state)
 {
-	struct psci_cpuidle_domain_state *ds = this_cpu_ptr(&psci_domain_state);
+	struct psci_cpuidle_data *data = this_cpu_ptr(&psci_cpuidle_data);
+	struct psci_cpuidle_domain_state *ds;
 
+	if (!data->collect_domain_states)
+		return 0;
+	if (data->domain_state_count >= data->domain_state_capacity)
+		return -ENOSPC;
+
+	ds = &data->domain_states[data->domain_state_count++];
 	ds->pd = pd;
 	ds->state_idx = state_idx;
 	ds->state = state;
+	return 0;
 }
 
 static inline void psci_clear_domain_state(void)
 {
-	__this_cpu_write(psci_domain_state.state, 0);
+	struct psci_cpuidle_data *data = this_cpu_ptr(&psci_cpuidle_data);
+
+	data->domain_state_count = 0;
+	data->collect_domain_states = false;
 }
 
 static __cpuidle int __psci_enter_domain_idle_state(struct cpuidle_device *dev,
@@ -69,22 +83,27 @@ static __cpuidle int __psci_enter_domain_idle_state(struct cpuidle_device *dev,
 	u32 *states = data->psci_states;
 	struct device *pd_dev = data->dev;
 	struct psci_cpuidle_domain_state *ds;
+	unsigned int i;
 	u32 state = states[idx];
 	int ret;
 
+	psci_clear_domain_state();
 	ret = cpu_pm_enter();
 	if (ret)
 		return -1;
 
 	/* Do runtime PM to manage a hierarchical CPU toplogy. */
+	data->collect_domain_states = true;
 	if (s2idle)
 		dev_pm_genpd_suspend_s2idle(pd_dev);
 	else
 		pm_runtime_put_sync_suspend(pd_dev);
+	data->collect_domain_states = false;
 
-	ds = this_cpu_ptr(&psci_domain_state);
-	if (ds->state)
+	if (data->domain_state_count) {
+		ds = &data->domain_states[data->domain_state_count - 1];
 		state = ds->state;
+	}
 
 	trace_psci_domain_idle_enter(dev->cpu, state, s2idle);
 	ret = psci_cpu_suspend_enter(state) ? -1 : idx;
@@ -98,8 +117,12 @@ static __cpuidle int __psci_enter_domain_idle_state(struct cpuidle_device *dev,
 	cpu_pm_exit();
 
 	/* Correct domain-idlestate statistics if we failed to enter. */
-	if (ret == -1 && ds->state)
-		pm_genpd_inc_rejected(ds->pd, ds->state_idx, s2idle);
+	if (ret == -1) {
+		for (i = 0; i < data->domain_state_count; i++) {
+			ds = &data->domain_states[i];
+			pm_genpd_inc_rejected(ds->pd, ds->state_idx, s2idle);
+		}
+	}
 
 	/* Clear the domain state to start fresh when back from idle. */
 	psci_clear_domain_state();
@@ -246,10 +269,29 @@ int psci_dt_parse_state_node(struct device_node *np, u32 *state)
 	return 0;
 }
 
-static int psci_dt_cpu_init_topology(struct cpuidle_driver *drv,
+static unsigned int psci_pd_count_domains(struct generic_pm_domain *pd)
+{
+	struct gpd_link *link;
+	unsigned int child_count;
+	unsigned int count = 1;
+
+	list_for_each_entry(link, &pd->child_links, child_node) {
+		child_count = psci_pd_count_domains(link->parent);
+		if (!child_count || check_add_overflow(count, child_count, &count))
+			return 0;
+	}
+
+	return count;
+}
+
+static int psci_dt_cpu_init_topology(struct device *dev,
+				     struct cpuidle_driver *drv,
 				     struct psci_cpuidle_data *data,
 				     unsigned int state_count, int cpu)
 {
+	struct generic_pm_domain *pd;
+	unsigned int domain_count;
+
 	/* Currently limit the hierarchical topology to be used in OSI mode. */
 	if (!psci_has_osi_support())
 		return 0;
@@ -257,6 +299,29 @@ static int psci_dt_cpu_init_topology(struct cpuidle_driver *drv,
 	data->dev = dt_idle_attach_cpu(cpu, "psci");
 	if (IS_ERR_OR_NULL(data->dev))
 		return PTR_ERR_OR_ZERO(data->dev);
+	if (!data->dev->pm_domain) {
+		dt_idle_detach_cpu(data->dev);
+		data->dev = NULL;
+		return -ENODEV;
+	}
+
+	pd = pd_to_genpd(data->dev->pm_domain);
+	domain_count = psci_pd_count_domains(pd);
+	if (!domain_count) {
+		dt_idle_detach_cpu(data->dev);
+		data->dev = NULL;
+		return -EOVERFLOW;
+	}
+
+	data->domain_states =
+		devm_kcalloc(dev, domain_count, sizeof(*data->domain_states),
+			     GFP_KERNEL);
+	if (!data->domain_states) {
+		dt_idle_detach_cpu(data->dev);
+		data->dev = NULL;
+		return -ENOMEM;
+	}
+	data->domain_state_capacity = domain_count;
 
 	psci_cpuidle_use_syscore = true;
 
@@ -306,7 +371,7 @@ static int psci_dt_cpu_init_idle(struct device *dev, struct cpuidle_driver *drv,
 		return -ENODEV;
 
 	/* Initialize optional data, used for the hierarchical topology. */
-	ret = psci_dt_cpu_init_topology(drv, data, state_count, cpu);
+	ret = psci_dt_cpu_init_topology(dev, drv, data, state_count, cpu);
 	if (ret < 0)
 		return ret;
 
@@ -344,6 +409,11 @@ static void psci_cpu_deinit_idle(int cpu)
 	struct psci_cpuidle_data *data = per_cpu_ptr(&psci_cpuidle_data, cpu);
 
 	dt_idle_detach_cpu(data->dev);
+	data->dev = NULL;
+	data->domain_states = NULL;
+	data->domain_state_count = 0;
+	data->domain_state_capacity = 0;
+	data->collect_domain_states = false;
 	psci_cpuidle_use_syscore = false;
 }
 
diff --git a/drivers/cpuidle/cpuidle-psci.h b/drivers/cpuidle/cpuidle-psci.h
index d29cbd796c..caf181247d 100644
--- a/drivers/cpuidle/cpuidle-psci.h
+++ b/drivers/cpuidle/cpuidle-psci.h
@@ -6,8 +6,8 @@
 struct device_node;
 struct generic_pm_domain;
 
-void psci_set_domain_state(struct generic_pm_domain *pd, unsigned int state_idx,
-			   u32 state);
+int psci_set_domain_state(struct generic_pm_domain *pd, unsigned int state_idx,
+			  u32 state);
 int psci_dt_parse_state_node(struct device_node *np, u32 *state);
 
 #endif /* __CPUIDLE_PSCI_H */
From 3c63503ce65b63d31001a13d9fd460182b635a7d Mon Sep 17 00:00:00 2001
From: Bharat Dintakurti <bdintakurti@nvidia.com>
Date: Tue, 1 Sep 2026 22:49:24 +0000
Subject: [PATCH] NVIDIA: SAUCE: ACPI: processor_idle: Clear freed cpuidle
 device pointers

BugLink: https://bugs.launchpad.net/bugs/2167301

acpi_processor_power_init() publishes each allocated cpuidle device through
the per-CPU acpi_cpuidle_device pointer. The normal processor power-exit
path and the global power-state refresh path unregister and free that
device without clearing the published pointer.

If reinitialization fails before installing a replacement, the per-CPU slot
continues to reference freed memory. A repeated or adjacent lifecycle
operation can then consume the dangling pointer.

Clear the owner slot immediately after unregistering the device and before
freeing it in both paths.

Fixes: e18afcb7b2a12 ("ACPI: processor_idle: Fix memory leak in acpi_processor_power_exit()")
Fixes: 07cba0de5598 ("ACPI: processor: idle: Reset cpuidle on C-state list changes")
Signed-off-by: Bharat Dintakurti <bdintakurti@nvidia.com>
Signed-off-by: Ishan Shah <ishah@nvidia.com>
Reviewed-by: Ishan Shah <ishah@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/acpi/processor_idle.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/acpi/processor_idle.c b/drivers/acpi/processor_idle.c
index 4482cf28f5..49c9f51e84 100644
--- a/drivers/acpi/processor_idle.c
+++ b/drivers/acpi/processor_idle.c
@@ -1319,6 +1319,7 @@ int acpi_processor_power_state_has_changed(struct acpi_processor *pr)
 				continue;
 
 			cpuidle_unregister_device_no_lock(dev);
+			per_cpu(acpi_cpuidle_device, cpu) = NULL;
 			kfree(dev);
 			_pr->flags.power = 0;
 		}
@@ -1461,6 +1462,7 @@ void acpi_processor_power_exit(struct acpi_processor *pr)
 
 	if (pr->flags.power) {
 		cpuidle_unregister_device(dev);
+		per_cpu(acpi_cpuidle_device, pr->id) = NULL;
 		kfree(dev);
 	}
 
From 8fb13cd71cb76faab1332bafebc3eb4152d2ede3 Mon Sep 17 00:00:00 2001
From: Ishan Shah <ishah@nvidia.com>
Date: Mon, 14 Sep 2026 22:53:27 +0000
Subject: [PATCH] NVIDIA: SAUCE: ACPI: processor_idle: Serialize idle-state
 rebuilds

BugLink: https://bugs.launchpad.net/bugs/2167301

A power-state notification rebuilds the shared ACPI idle driver and every
cpuidle device.  The cpuidle lock protects individual registration
operations, but it is released between device teardown, driver replacement,
and device recreation.

Concurrent CPU0 notifications can therefore interleave complete rebuild
transactions and lose or free each other's objects.  Serialize the
system-wide transaction with a dedicated mutex.  Recheck the processor
state after taking the mutex because another rebuild may have completed
while the notification waited.

Signed-off-by: Ishan Shah <ishah@nvidia.com>
Assisted-by: OpenAI-Codex:gpt-5.6
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/acpi/processor_idle.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/drivers/acpi/processor_idle.c b/drivers/acpi/processor_idle.c
index 49c9f51e84..42e0e08372 100644
--- a/drivers/acpi/processor_idle.c
+++ b/drivers/acpi/processor_idle.c
@@ -50,6 +50,7 @@ static unsigned int latency_factor __read_mostly = 2;
 module_param(latency_factor, uint, 0644);
 
 static DEFINE_PER_CPU(struct cpuidle_device *, acpi_cpuidle_device);
+static DEFINE_MUTEX(acpi_idle_rebuild_lock);
 
 static struct cpuidle_driver acpi_idle_driver = {
 	.name =		"acpi_idle",
@@ -1291,14 +1292,20 @@ int acpi_processor_hotplug(struct acpi_processor *pr)
 int acpi_processor_power_state_has_changed(struct acpi_processor *pr)
 {
 	int cpu;
+	int ret = 0;
 	struct acpi_processor *_pr;
 	struct cpuidle_device *dev;
 
 	if (disabled_by_idle_boot_param())
 		return 0;
+	if (pr->id != 0)
+		return 0;
 
-	if (!pr->flags.power_setup_done)
-		return -ENODEV;
+	mutex_lock(&acpi_idle_rebuild_lock);
+	if (!pr->flags.power_setup_done) {
+		ret = -ENODEV;
+		goto out;
+	}
 
 	/*
 	 * FIXME:  Design the ACPI notification to make it once per
@@ -1306,7 +1313,7 @@ int acpi_processor_power_state_has_changed(struct acpi_processor *pr)
 	 * to make the code that updates C-States be called once.
 	 */
 
-	if (pr->id == 0 && cpuidle_get_driver() == &acpi_idle_driver) {
+	if (cpuidle_get_driver() == &acpi_idle_driver) {
 		/* Protect against cpu-hotplug */
 		cpus_read_lock();
 
@@ -1347,7 +1354,9 @@ int acpi_processor_power_state_has_changed(struct acpi_processor *pr)
 		cpus_read_unlock();
 	}
 
-	return 0;
+out:
+	mutex_unlock(&acpi_idle_rebuild_lock);
+	return ret;
 }
 
 void acpi_processor_register_idle_driver(void)
